content/en/blog/_posts/2021-08-11-memory-qos-cgroups-v2.md

```diff
@@ -0,0 +1,73 @@
+---
+layout: blog
+title: 'Support memory qos with cgroups v2'
+date: 2021-08-11
+slug: memory-qos-cgroups-v2
+---
+
+**Authors:** Tim Xu (Tencent Cloud)
+
+Memory QoS is used to memory protection and usage throttle on pod / container to increase memory isolation and service quality.
+
+In traditional cgroups v1 in Kubernetes, cpu resources can be very well protected. Kubernetes uses `cpu_shares / cpu_set / cpu_quota / cpu_period` to isolate cpu. But the memory is not well guaranteed, especially at cgroup level. Kubernetes only uses `memory.limit_in_bytes=sum(pod.spec.containers.resources.limits[memory])` with cgroups v1 and `memory.max=sum(pod.spec.containers.resources.limits[memory])` with cgroups v2 to limit memory usage. `resources.requests[memory]` is not yet used neither by cgroups v1 nor cgroups v2 to protect memory requests. About memory protection, we use `oom_scores` to determine order of killing container process when OOM occurs.
+
+It may cause:
+- Pod / Container memory requests can't be fully reserved, page cache is at risk of being recycled
+- Pod / Container memory allocation is not well protected, there may occur allocation latency frequently when node memory nearly runs out
+- Memory overcommit of container is not throttled, there may increase risk of node memory pressure
+
+Kubernetes 1.22 brings Memory QoS to alpha to pay attention on the quality of service of memory resources. It uses cgroup v2 new capabilities `memory.min / memory.high` in memory controller to retain memory requests and throttle overcommit limits. This feature can improve memory availability and reduce memory requests reclamation, which is of great significance to ensure the stability of both workloads and nodes.
```

Comment:
```suggestion
Kubernetes 1.22 brings Memory QoS to alpha to pay attention to the quality of service of memory resources. It uses cgroup v2 new capabilities `memory.min` and `memory.high` in memory controller to retain memory requests and throttle overcommit limits. This feature can improve memory availability and reduce memory requests reclamation, which is of great significance to ensure the stability of both workloads and nodes.
```

Reply:
Is this correct:
```suggestion
Kubernetes 1.22 introduces a new alpha feature for defining quality-of-service for memory
resources. This new feature relies on new cgroup v2 capabilities: `memory.min` and
`memory.high`. The kubelet's memory management subsystem uses those capabilities to
define a guaranteed amount of RAM for the cgroup overall, as well as to restrict how much
memory that control group can try to overcommit. Using memory QoS on your nodes can
help you to improve application availability. The `memory.min` setting also helps avoid
paging related to the Pods and containers that it applies to, which is important for
latency-sensitive workloads.
```
?

Reply:
It's basically right. I made some minor adjustments.